File: src/main/java/me/anonymoussoftware/vacancymanager/VacancyManager.java
package me.anonymoussoftware.vacancymanager;

import java.io.File;
import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.TreeMap;
import java.util.concurrent.CopyOnWriteArrayList;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.stream.Collectors;

import org.springframework.beans.factory.DisposableBean;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import me.anonymoussoftware.vacancymanager.api.service.FileVacancyService;
import me.anonymoussoftware.vacancymanager.api.service.HttpVacancyService;
import me.anonymoussoftware.vacancymanager.model.Employer;
import me.anonymoussoftware.vacancymanager.model.Vacancy;
import me.anonymoussoftware.vacancymanager.model.aggregated.AggregatedVacancy;

@Service
public class VacancyManager implements DisposableBean {

    private static final int VACANCY_COUNT_THRESHOLD = 2000;

    @Autowired
    private FileVacancyService fileVacancyService;

    @Autowired
    private HttpVacancyService httpVacancyService;

    private final ExecutorService threadPool = Executors.newSingleThreadExecutor();

    private final List<VacancyListChangeListener> vacancyChangeListeners = new CopyOnWriteArrayList<>();

    private final List<VacancySelectionListener> vacancySelectionListeners = new CopyOnWriteArrayList<>();

    private final List<VacancySearchListener> vacancySearchListeners = new CopyOnWriteArrayList<>();

    private final List<EmployerListChangeListener> employerListChangeListeners = new CopyOnWriteArrayList<>();

    private int selectedCityCode;

    private AggregatedVacancy selectedVacancy;

    private List<Vacancy> vacancies = new ArrayList<>();

    private Map<Integer, Employer> employers = new TreeMap<>();

    private final Map<String, Integer> CITIES = new LinkedHashMap<>();

    {
        CITIES.put("Saint Petersburg", 2);
        CITIES.put("Moscow", 1);
    }

    public void addVacancyListChangeListener(VacancyListChangeListener listener) {
        this.vacancyChangeListeners.add(listener);
    }

    public void removeVacancyListChangeListener(VacancyListChangeListener listener) {
        this.vacancyChangeListeners.remove(listener);
    }

    public void addVacancySelectionListener(VacancySelectionListener listener) {
        this.vacancySelectionListeners.add(listener);
    }

    public void removeVacancySelectionListener(VacancySelectionListener listener) {
        this.vacancySelectionListeners.remove(listener);
    }

    public void addVacancySearchListener(VacancySearchListener listener) {
        this.vacancySearchListeners.add(listener);
    }

    public void removeVacancySearchListener(VacancySearchListener listener) {
        this.vacancySearchListeners.remove(listener);
    }

    public void addEmployerListChangeListener(EmployerListChangeListener listener) {
        this.employerListChangeListeners.add(listener);
    }

    public void removeEmployerListChangeListener(EmployerListChangeListener listener) {
        this.employerListChangeListeners.remove(listener);
    }

    public Set<String> getAvailableCityNames() {
        return Collections.unmodifiableSet(CITIES.keySet());
    }

    public void selectCity(String cityName) {
        Integer selectedCityTmp = CITIES.get(cityName);
        if (selectedCityTmp == null) {
            throw new IllegalStateException("Unknown city \"" + cityName + "\"");
        }
        this.selectedCityCode = selectedCityTmp;
    }

    public void banSelectedVacancy() {
        this.selectedVacancy.getVacancy().setBanned(true);
        fireVacancyListChanged(VacancyListChangeListener.VacancyListChangeReason.VACANCY_BAN);
    }

    public void banSelectedVacancyEmployer() {
        Employer employerToBan = this.selectedVacancy.getVacancy().getEmployer();
        this.employers.get(employerToBan.getId()).setBanned(true);
        fireVacancyListChanged(VacancyListChangeListener.VacancyListChangeReason.VACANCY_BAN);
        fireEmployerListChanged();
    }

    public int getSelectedCityCode() {
        return this.selectedCityCode;
    }

    private void loadVacancies(List<Vacancy> vacancies) {
        this.vacancies = vacancies;
        fireVacancyListChanged(VacancyListChangeListener.VacancyListChangeReason.TOTAL_RELOAD);
        fireEmployerListChanged();
    }

    private void fireVacancyListChanged(VacancyListChangeListener.VacancyListChangeReason reason) {
        this.vacancyChangeListeners.stream().forEach(listener -> listener.onVacancyListChange(reason));
    }

    private void fireEmployerListChanged() {
        this.employerListChangeListeners.stream().forEach(EmployerListChangeListener::onEmployerListChange);
    }

    public void startVacanciesSearch(String searchText) {
        this.threadPool.submit(() -> {
            List<Vacancy> vacancies = new ArrayList<>();
            int currentPage = 0;
            int total = 0;
            int vacanciesToLoad = 0;
            fireVacancySearchProgress(0);
            Set<Integer> bannedEmployers = this.employers.values().stream() //
                    .filter(Employer::isBanned) //
                    .map(Employer::getId) //
                    .collect(Collectors.toSet());
            do {
                String url = this.httpVacancyService.getRequestVacancyUrl(searchText, getSelectedCityCode(),
                        currentPage, bannedEmployers);
                fireVacancyRequestPageStart(url);
                HttpVacancyService.VacancyListResult result = this.httpVacancyService.requestVacancies(url);
                vacancies.addAll(cacheVacancies(result.getVacancies()));
                total = result.getTotal();
                vacanciesToLoad = Math.min(total, VACANCY_COUNT_THRESHOLD);
                currentPage++;
                fireVacancySearchProgress((int) ((vacancies.size() + 0.0) / vacanciesToLoad * 100));
            } while (vacancies.size() < vacanciesToLoad);
            loadVacancies(vacancies);
            fireVacancySearchProgress(100);
            fireVacancySearchFinish(vacancies.size(), total);
        });
    }

    private void fireVacancySearchFinish(int loaded, int total) {
        for (VacancySearchListener vacancySearchListener : this.vacancySearchListeners) {
            vacancySearchListener.onSearchFinish(loaded, total);
        }
    }

    private void fireVacancySearchProgress(int percentage) {
        for (VacancySearchListener vacancySearchListener : this.vacancySearchListeners) {
            vacancySearchListener.onVacancyPageSuccessfulLoad(percentage);
        }
    }

    private void fireVacancyRequestPageStart(String url) {
        for (VacancySearchListener vacancySearchListener : this.vacancySearchListeners) {
            vacancySearchListener.onStartVacancyPageLoad(url);
        }
    }

    public List<AggregatedVacancy> getAvailableVacancies() {
        Map<Integer, List<Vacancy>> employerToVacancies = this.vacancies.stream()
                .collect(Collectors.groupingBy((Vacancy v) -> v.getEmployer().getId()));
        return this.vacancies.stream()
                .map(v -> new AggregatedVacancy(v, employerToVacancies.get(v.getEmployer().getId()))) //
                .sorted(Comparator.comparing((AggregatedVacancy v) -> v.getVacancy().isBanned()) //
                        .thenComparing(v -> v.getVacancy().getEmployer().isBanned()) //
                        .thenComparing(v -> -v.getEmployerVacancies().size()) //
                        .thenComparing(v -> v.getVacancy().getEmployer().getName()) //
                        .thenComparing(v -> v.getVacancy().getName()))
                .collect(Collectors.toList());
    }

    public List<Employer> getEmployers() {
        return this.employers.values().stream() //
                .sorted(Comparator.comparing((Employer e) -> !e.isBanned()) //
                        .thenComparing(Employer::getName))
                .collect(Collectors.toList());
    }

    public void setSelectedVacancy(AggregatedVacancy vacancy) {
        this.selectedVacancy = vacancy;
        this.vacancySelectionListeners.stream().forEach(e -> e.onVacancySelection());
    }

    public AggregatedVacancy getSelectedVacancy() {
        return this.selectedVacancy;
    }

    public void importVacancies(File file, Runnable successAction, Runnable failAction) {
        boolean success = this.fileVacancyService.importVacancies(this.vacancies, file);
        if (success) {
            successAction.run();
        } else {
            failAction.run();
        }
    }

    private List<Vacancy> cacheVacancies(List<Vacancy> rawVacancies) {
        List<Vacancy> newVacancies = new ArrayList<>();
        for (Vacancy rawVacancy : rawVacancies) {
            Employer rawEmployer = rawVacancy.getEmployer();
            Employer cachedEmployer = this.employers.get(rawEmployer.getId());
            if (cachedEmployer == null) {
                this.employers.put(rawEmployer.getId(), rawEmployer);
                cachedEmployer = rawEmployer;
            }
            newVacancies.add(new Vacancy(rawVacancy.getId(), rawVacancy.isBanned(), rawVacancy.getName(),
                    cachedEmployer, rawVacancy.getArea(), rawVacancy.getSnippet()));
        }
        return newVacancies;
    }

    private void updateEmployersCache(List<Employer> bannedEmployers) {
        for (Employer bannedEmployer : bannedEmployers) {
            Employer cachedEmployer = this.employers.get(bannedEmployer.getId());
            if (cachedEmployer == null) {
                this.employers.put(bannedEmployer.getId(), bannedEmployer);
            } else {
                cachedEmployer.setBanned(true);
            }
        }
        fireVacancyListChanged(VacancyListChangeListener.VacancyListChangeReason.TOTAL_RELOAD);
        fireEmployerListChanged();
    }

    public void exportVacancies(File file, Runnable failAction) {
        List<Vacancy> nonCachedVacancies = this.fileVacancyService.exportVacancies(file);
        if (nonCachedVacancies != null) {
            List<Vacancy> cachedResult = cacheVacancies(nonCachedVacancies);
            loadVacancies(cachedResult);
        } else {
            failAction.run();
        }
    }

    public void importEmployers(File file, Runnable successAction, Runnable failAction) {
        boolean success = this.fileVacancyService.importBannedEmployers(this.employers.values(), file);
        if (success) {
            successAction.run();
        } else {
            failAction.run();
        }
    }

    public void exportEmployers(File file, Runnable failAction) {
        List<Employer> nonCachedVacancies = this.fileVacancyService.exportBannedEmployers(file);
        if (nonCachedVacancies != null) {
            updateEmployersCache(nonCachedVacancies);
        } else {
            failAction.run();
        }
    }

    public static interface VacancySearchListener {
        void onVacancyPageSuccessfulLoad(int percentage);

        void onStartVacancyPageLoad(String url);

        void onSearchFinish(int loaded, int total);
    }

    public static interface VacancyListChangeListener {
        public enum VacancyListChangeReason {
            TOTAL_RELOAD, VACANCY_BAN
        }

        void onVacancyListChange(VacancyListChangeReason reason);
    }

    public static interface EmployerListChangeListener {
        void onEmployerListChange();
    }

    public static interface VacancySelectionListener {
        void onVacancySelection();
    }

    @Override
    public void destroy() throws Exception {
        this.threadPool.shutdown();
    }
}

File: src/main/kotlin/me/anonymoussoftware/vacancymanager/VacancyManager.kt
package me.anonymoussoftware.vacancymanager

import me.anonymoussoftware.vacancymanager.api.service.FileVacancyService
import me.anonymoussoftware.vacancymanager.api.service.HttpVacancyService
import me.anonymoussoftware.vacancymanager.model.Employer
import me.anonymoussoftware.vacancymanager.model.Vacancy
import me.anonymoussoftware.vacancymanager.model.aggregated.AggregatedVacancy
import org.springframework.beans.factory.DisposableBean
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Service
import java.io.File
import java.util.*
import java.util.concurrent.CopyOnWriteArrayList
import java.util.concurrent.Executors
import java.util.function.Consumer
import java.util.function.Predicate
import java.util.stream.Collectors

@Service
class VacancyManager : DisposableBean {

    @Autowired
    private val fileVacancyService: FileVacancyService? = null

    @Autowired
    private val httpVacancyService: HttpVacancyService? = null

    private val threadPool = Executors.newSingleThreadExecutor()

    private val vacancyChangeListeners = CopyOnWriteArrayList<VacancyListChangeListener>()

    private val vacancySelectionListeners = CopyOnWriteArrayList<VacancySelectionListener>()

    private val vacancySearchListeners = CopyOnWriteArrayList<VacancySearchListener>()

    private val employerListChangeListeners = CopyOnWriteArrayList<EmployerListChangeListener>()

    private val vacancyDescriptionRequestListeners = CopyOnWriteArrayList<VacancyDescriptionRequestListener>()

    var selectedCityCode: Int = 0
        private set

    var selectedVacancy: AggregatedVacancy? = null
        set(vacancy) {
            field = vacancy
            this.vacancySelectionListeners.stream().forEach { e -> e.onVacancySelection() }
        }

    private var vacancies: List<Vacancy> = ArrayList()

    private val employers = TreeMap<Int, Employer>()

    private val CITIES = LinkedHashMap<String, Int>()

    val availableCityNames: Set<String>
        get() = Collections.unmodifiableSet(CITIES.keys)

    //
    //
    //
    //
    //
    val availableVacancies: List<AggregatedVacancy>
        get() {
            val employerToVacancies = this.vacancies.stream()
                .collect(Collectors.groupingBy { v: Vacancy -> v.employer!!.id })
            return this.vacancies.stream()
                .map { v -> AggregatedVacancy(v, employerToVacancies[v.employer!!.id]) }
                .sorted(Comparator.comparing { v: AggregatedVacancy -> v.vacancy!!.isBanned }
                    .thenComparing { v -> v.vacancy!!.employer!!.isBanned }
                    .thenComparing { v -> -v.employerVacancies!!.size }
                    .thenComparing { v -> v.vacancy!!.employer!!.name!! }
                    .thenComparing { v -> v.vacancy!!.name!! })
                .collect(Collectors.toList())
        }

    init {
        CITIES["Saint Petersburg"] = 2
        CITIES["Moscow"] = 1
        CITIES["Sochi"] = 237
        CITIES["Krasnodar"] = 53
        CITIES["Chelyabinsk"] = 104
    }

    fun addVacancyListChangeListener(listener: VacancyListChangeListener) {
        this.vacancyChangeListeners.add(listener)
    }

    fun removeVacancyListChangeListener(listener: VacancyListChangeListener) {
        this.vacancyChangeListeners.remove(listener)
    }

    fun addVacancySelectionListener(listener: VacancySelectionListener) {
        this.vacancySelectionListeners.add(listener)
    }

    fun removeVacancySelectionListener(listener: VacancySelectionListener) {
        this.vacancySelectionListeners.remove(listener)
    }

    fun addVacancySearchListener(listener: VacancySearchListener) {
        this.vacancySearchListeners.add(listener)
    }

    fun removeVacancySearchListener(listener: VacancySearchListener) {
        this.vacancySearchListeners.remove(listener)
    }

    fun addEmployerListChangeListener(listener: EmployerListChangeListener) {
        this.employerListChangeListeners.add(listener)
    }

    fun removeEmployerListChangeListener(listener: EmployerListChangeListener) {
        this.employerListChangeListeners.remove(listener)
    }

    fun addVacancyDescriptionRequestListener(listener: VacancyDescriptionRequestListener) {
        this.vacancyDescriptionRequestListeners.add(listener)
    }

    fun removeVacancyDescriptionRequestListener(listener: VacancyDescriptionRequestListener) {
        this.vacancyDescriptionRequestListeners.remove(listener)
    }

    fun selectCity(cityName: String) {
        val selectedCityTmp = CITIES[cityName] ?: throw IllegalStateException("Unknown city \"$cityName\"")
        this.selectedCityCode = selectedCityTmp
    }

    fun banSelectedVacancy() {
        this.selectedVacancy!!.vacancy!!.isBanned = true
        fireVacancyListChanged(VacancyListChangeListener.VacancyListChangeReason.VACANCY_BAN)
    }

    fun banSelectedVacancyEmployer() {
        val employerToBan = this.selectedVacancy!!.vacancy!!.employer
        this.employers[employerToBan!!.id]?.isBanned = true
        fireVacancyListChanged(VacancyListChangeListener.VacancyListChangeReason.VACANCY_BAN)
        fireEmployerListChanged()
    }

    fun requestSelectedVacancyDescription() {
        if (this.selectedVacancy == null) {
            return
        }
        this.threadPool.submit {
            val vacancy = this.selectedVacancy!!.vacancy
            val id = vacancy!!.id
            val url = this.httpVacancyService!!.getRequestVacancyDescriptionUrl(id)
            fireVacancyDescriptionRequestStart(url)
            val description = this.httpVacancyService.requestVacancyDescription(url)
            vacancy.description = description
            fireVacancyDescriptionRequestFinish()
        }
    }

    private fun loadVacancies(vacancies: List<Vacancy>) {
        this.vacancies = vacancies
        fireVacancyListChanged(VacancyListChangeListener.VacancyListChangeReason.TOTAL_RELOAD)
        fireEmployerListChanged()
    }

    private fun fireVacancyListChanged(reason: VacancyListChangeListener.VacancyListChangeReason) {
        this.vacancyChangeListeners.stream().forEach { listener -> listener.onVacancyListChange(reason) }
    }

    private fun fireVacancyDescriptionRequestStart(url: String) {
        this.vacancyDescriptionRequestListeners.stream()
            .forEach { listener -> listener.onVacancyDescriptionRequestStart(url) }
    }

    private fun fireVacancyDescriptionRequestFinish() {
        this.vacancyDescriptionRequestListeners.stream()
            .forEach { listener -> listener.onVacancyDescriptionRequestFinish() }
    }

    private fun fireEmployerListChanged() {
        this.employerListChangeListeners.stream()
            .forEach(Consumer<EmployerListChangeListener> { it.onEmployerListChange() })
    }

    fun startVacanciesSearch(searchText: String) {
        this.threadPool.submit {
            val vacancies = ArrayList<Vacancy>()
            var currentPage = 0
            var total = 0
            var vacanciesToLoad = 0
            fireVacancySearchProgress(0)
            val bannedEmployers = this.employers.values.stream() //
                .filter(Predicate<Employer> { it.isBanned }) //
                .map { it.id } //
                .collect(Collectors.toSet())
            do {
                val url = this.httpVacancyService!!.getRequestVacancyUrl(
                    searchText, selectedCityCode,
                    currentPage, bannedEmployers
                )
                fireVacancyRequestPageStart(url)
                val result = this.httpVacancyService.requestVacancies(url)
                vacancies.addAll(cacheVacancies(result!!.vacancies))
                total = result.total
                vacanciesToLoad = Math.min(total, VACANCY_COUNT_THRESHOLD)
                currentPage++
                fireVacancySearchProgress(((vacancies.size + 0.0) / vacanciesToLoad * 100).toInt())
            } while (vacancies.size < vacanciesToLoad)
            loadVacancies(vacancies)
            fireVacancySearchProgress(100)
            fireVacancySearchFinish(vacancies.size, total)
        }
    }

    private fun fireVacancySearchFinish(loaded: Int, total: Int) {
        for (vacancySearchListener in this.vacancySearchListeners) {
            vacancySearchListener.onSearchFinish(loaded, total)
        }
    }

    private fun fireVacancySearchProgress(percentage: Int) {
        for (vacancySearchListener in this.vacancySearchListeners) {
            vacancySearchListener.onVacancyPageSuccessfulLoad(percentage)
        }
    }

    private fun fireVacancyRequestPageStart(url: String) {
        for (vacancySearchListener in this.vacancySearchListeners) {
            vacancySearchListener.onStartVacancyPageLoad(url)
        }
    }

    fun getEmployers(): List<Employer> {
        return this.employers.values.stream() //
            .sorted(Comparator.comparing { e: Employer -> !e.isBanned } //
                .thenComparing { e: Employer -> e.name!! })
            .collect(Collectors.toList())
    }

    fun importVacancies(file: File, successAction: Runnable, failAction: Runnable) {
        val success = this.fileVacancyService!!.importVacancies(this.vacancies, file)
        if (success) {
            successAction.run()
        } else {
            failAction.run()
        }
    }

    private fun cacheVacancies(rawVacancies: List<Vacancy>): List<Vacancy> {
        val newVacancies = ArrayList<Vacancy>()
        for (rawVacancy in rawVacancies) {
            val rawEmployer = rawVacancy.employer
            var cachedEmployer: Employer? = this.employers[rawEmployer!!.id]
            if (cachedEmployer == null) {
                this.employers[rawEmployer!!.id] = rawEmployer
                cachedEmployer = rawEmployer
            }
            newVacancies.add(
                Vacancy(
                    rawVacancy.id, //
                    rawVacancy.isBanned, //
                    rawVacancy.name, //
                    cachedEmployer, //
                    rawVacancy.area, //
                    rawVacancy.snippet, //
                    rawVacancy.salary, //
                    rawVacancy.url, //
                    rawVacancy.description
                )
            )
        }
        return newVacancies
    }

    private fun updateEmployersCache(bannedEmployers: List<Employer>) {
        for (bannedEmployer in bannedEmployers) {
            val cachedEmployer = this.employers[bannedEmployer.id]
            if (cachedEmployer == null) {
                this.employers[bannedEmployer.id] = bannedEmployer
            } else {
                cachedEmployer.isBanned = true
            }
        }
        fireVacancyListChanged(VacancyListChangeListener.VacancyListChangeReason.TOTAL_RELOAD)
        fireEmployerListChanged()
    }

    fun exportVacancies(file: File, failAction: Runnable) {
        val nonCachedVacancies = this.fileVacancyService!!.exportVacancies(file)
        if (nonCachedVacancies != null) {
            val cachedResult = cacheVacancies(nonCachedVacancies)
            loadVacancies(cachedResult)
        } else {
            failAction.run()
        }
    }

    fun importEmployers(file: File, successAction: Runnable, failAction: Runnable) {
        val success = this.fileVacancyService!!.importBannedEmployers(this.employers.values, file)
        if (success) {
            successAction.run()
        } else {
            failAction.run()
        }
    }

    fun exportEmployers(file: File, failAction: Runnable) {
        val nonCachedVacancies = this.fileVacancyService!!.exportBannedEmployers(file)
        if (nonCachedVacancies != null) {
            updateEmployersCache(nonCachedVacancies)
        } else {
            failAction.run()
        }
    }

    interface VacancySearchListener {
        fun onVacancyPageSuccessfulLoad(percentage: Int)

        fun onStartVacancyPageLoad(url: String)

        fun onSearchFinish(loaded: Int, total: Int)
    }

    interface VacancyListChangeListener {
        enum class VacancyListChangeReason {
            TOTAL_RELOAD, VACANCY_BAN
        }

        fun onVacancyListChange(reason: VacancyListChangeReason)
    }

    interface EmployerListChangeListener {
        fun onEmployerListChange()
    }

    interface VacancySelectionListener {
        fun onVacancySelection()
    }

    interface VacancyDescriptionRequestListener {
        fun onVacancyDescriptionRequestStart(url: String)

        fun onVacancyDescriptionRequestFinish()
    }

    @Throws(Exception::class)
    override fun destroy() {
        this.threadPool.shutdown()
    }

    companion object {

        private val VACANCY_COUNT_THRESHOLD = 2000
    }
}
